src/tests/test_dal_middleware.py

```diff
@@ -0,0 +1,303 @@
+import pytest
+
+from polydatum import DataAccessLayer, Service
+from polydatum.dal import DalMethodResolverMiddleware, MethodMiddleware, InvalidMiddleware, DalMethodRequest, \
+    default_handle_dal_method, DataManager, PathSegment, DalMethodRequester
+
+
+def test_dal_middleware_requires_method_middleware():
+    """
+    Verify that the DAL will raise an error if you try to
+    register method middleware that does not subclass the
+    base class for method middleware: MethodMiddleware.
+    """
+    class FooMiddleware:
+        pass
+
+    with pytest.raises(InvalidMiddleware):
+        DataAccessLayer(data_manager=None, middleware=[FooMiddleware()])
+
+
+def test_dal_middleware_instantiation_not_needed():
+    """
+    Verify that you can specify method middleware as either an instance
+    or a class (as a convenience).
+    """
+    class FooMiddleware(MethodMiddleware):
+        pass
+
+    dal = DataAccessLayer(data_manager=None, middleware=[FooMiddleware], default_middleware=None)
+    assert isinstance(dal._reversed_middleware[0], FooMiddleware)
+
+    dal = DataAccessLayer(data_manager=None, middleware=[FooMiddleware()], default_middleware=None)
+    assert isinstance(dal._reversed_middleware[0], FooMiddleware)
+
+
+def test_dal_middleware_custom_middleware():
+    """
+    Verify determining all middleware for the DAL to manage will
+    use our default middleware if none is specified, AND if any is
+    specified, our default middleware will be added as the last
+    middleware.
+
+    We also need to verify that these middlewares chain each other
+    in the order that we expect (reversed) and that they ultimately
+    use the default handler for dal methods.
+    """
+    # Verify default behavior first.
+    # Passing an empty DM here because we don't care about this
+    # for the purposes of the test, and it's not used.
+    dal = DataAccessLayer(data_manager=None)
+    assert len(dal._reversed_middleware) == 1
```

Comment:
Poking into private properties to verify tests is a code smell. If it's useful to verify in tests, it's probably either useful outside of tests or some other input/output should be verified indirectly.

`test_dal_method_middleware` seems to cover checking this beautifully without using private properties.

Reply:
I need a suggestion for how to do this. I explained in another comment that we literlaly only use this property for tests. I'd rather the property not exist, but you can't return anything from an `__init__` so I lack the cleverness needed to verify the order here (or I'm being too lazy to inspect the callable chains).

> test_dal_method_middleware seems to cover checking this beautifully without using private properties.

I wanted to be very thorough about asserting this ordering, I'd rather leave this test in since it's so explicit.

Reply:
I think you can do it just with the test below. 

```python
assert middleware_requests == [
'outer-middleware-ingress',
'second-middleware-ingress',
'default-middleware-ingress',
'default-middleware-outgress',
'second-middleware-outgress',
'outer-middleware-outgress',
]
```

This tells you everything is in the order you want it in without actually checking the order of middleware itself. We care about input/output in tests -- not the means of how it was achieved.